Pull Request for 11 Jan 2020.

jjj0611 opened the pull request.

jjj0611 committed:
Revert "Add all codes which were written in the saturday morning"

This reverts commit 82fc8e883bc6654345e94ceefb79481052cfd932.

to start with TDD

jjj0611 committed:
Add feature list in README.md

jjj0611 committed:
Complete feature #1

	- made Car class with 2 instance variables.
	- made get methods for them.
	- throw exception with constructing object with String over 5 leghth.

jjj0611 committed:
Complete feature #2.

	- add move method with 1 argument.
	- position will be added up with 1 if parameter is over 3.

jjj0611 committed:
Add exception case with constructing Car object.

	- throw exception when constructing with empty String.

jjj0611 committed:
Add feature list in READEME.md

jjj0611 committed:
Complete feature #3.

	- generate Car list from String list
	- validate duplication
	- validate number of cars

jjj0611 committed:
Complete feature #4.

	- make Trial class
	- validate if String is a number
	- validate if number is not below 1

jjj0611 committed:
Complete feature #5.

	- make RandomNumberGenerator with static method

jjj0611 committed:
Complete feature #8.

	- make CarSnapShot with overriding toString
	- make method getCarSnapShot from Car
	- make StringUtils with multiplying string.

jjj0611 committed:
Override equals and hashcode in Car

	- override equals and hashcode methods in Car beacuase of validating duplication
	- refactoring RacingCar Test.

jjj0611 committed:
Complete feature 11.

	- add GameInitializer to input from user

jjj0611 committed:
Modify Car and Trial

	- constructing Car with String name, add name.trim() to remove the empty space.
	- change Trial constructor parameter from String to int.

jjj0611 committed:
Complete feature 11.

	- create GameInputScanner with two method.
	- input car name from user
	- input trial times from user

jjj0611 committed:
Made InvalidInputException.

	- made an exception about invalid input.
	- extends IllegalArgumentException.

jjj0611 committed:
Refactoring all new constructor and exception.

	- made static factory method "from" in every class.
	- change all IllegalArgumentException to InvalidInputException.

jjj0611 committed:
Reafoactor all constructor & exception

	- made static factory method 'from' to replace constructor.
	- change all IllegalArugmentException to InvalidInputException.

jjj0611 committed:
Merge branch '장재주' of https://github.com/jjj0611/java-racingcar into 장재주

jjj0611 committed:
organize package directory

jjj0611 committed:
chnage access modifier on CarSnapShot static factory method 'from'

jjj0611 committed:
change moveforward method in car

	- argument is changed to CarProceedStrategy interface
	- make AlwaysProceedStrategy & AlwaysNotProceedStrategy
	- make RandomCarProceedStrategy

dain7672 commented on src/main/java/domain/Car.java:
```diff
@@ -1,36 +1,62 @@
-    public void goOrStop() {
-        int randomNumber = (int) (Math.random() * 10);
-        if (randomNumber > 4) {
+    public void moveForward(int number) {
```
원래의 이름인 goOrStop이 이 함수에는 더 잘 어울리는 것 같아요. (moveOrNot 이라거나..)
호출하는 쪽에서
car.moveForward(5)라고 했을 때는, 코드를 열어보지 않고서야, '5번 앞으로' 라고 생각하게 될 것 같아요.

dain7672 replied:
이름을 고민하면서 객체나 함수의 정체성을 더 정확히 알게 되기도 해요. goOrStop도 사실 좋은 이름은 아닌데요. 정 고민해도 안좋은 이름이 나온다, 그럼 클래스의 역할과 설계에 대해서 다시 생각해볼 시간이에요.
차 스스로 앞으로 갈지 말지 판단하게 해줘도 되는 걸까? 등등

jjj0611 replied:
자동차가 어떻게 가게 만들까?라는 것에서
1. 자동차 내부에서 알아서 난수를 생성해서 판단한다.
2. 자동차가 특정한 것을 인자로 받아서 결정한다.
3. 자동차는 무조건 가기만 하고, 외부에서 결정한다.

이렇게 3가지를 고민해봤는데
1번은 자동차 부터 테스트 하기가 어려워지고 -> 결국 다 테스트 하기 어려워진다.
3번은 자동차 객체가 본인이 갈지 말지를 다른 누군가가 판단하게 하는게 옳을까?
이런 생각으로 2번으로 선택을 했었는데,
3번 처럼 구현을 했어도 괜찮을거 같네요.

dain7672 commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Test
+    void testCar() {
+        assertNotNull(car.getName());
```
assertEqual로 비교를 하는 게 더 좋겠죠?
큰 비용이 들지 않는다면 더 정확한 테스트를 작성하는 게 
원래 테스트를 하고자 하는 목적에 더 부합할 거에요.

dain7672 commented on src/test/java/domain/RacingCarsTest.java:
```diff
@@ -0,0 +1,49 @@
+        aloneCarNameList.add(carName);
+        List<Car> aloneCarList = aloneCarNameList.stream().map(Car::from).collect(Collectors.toList());
+        assertThrows(IllegalArgumentException.class, () -> RacingCars.from(aloneCarList));
+    }
```
1. (성공) 이름을 받아서 자동차를 생성한다.
2. (실패) 중복이 있으면 에러
3. (실패) 1개 이하면 에러

케이스 별로 테스트코드 메소드를 분리하게 되면 
- 정확한 테스트: 요구사항들에 대한 테스트를 빠짐없이 진행할 수 있구요,
- 유지보수:  요구사항이 변경되었을 때 테스트코드 수정이 용이해요. 
- 의미전달:  테스트코드는 테스트의 역할도 있지만, 일종의 명세서 역할을 하기도 해요. 요구사항을 담고 있는 거죠. 다른 사람이 이 클래스의 기능을 알고 싶을 때 잘 짜여진 테스트코드를 보고 쉽게 이해할 수 있어요.

jjj0611 replied:
앗 그러네요.
테스트를 위한 상황 3개를 한 곳에서 테스트를 하고 있었네요
분리 했습니다 :)

dain7672 commented on src/main/java/domain/Car.java:
```diff
@@ -1,36 +1,62 @@
+    public boolean equals(Object o) {
+        if (!(o instanceof Car))
+            return false;
+        return ((Car) o).getName().equals(name);
```
o.getName().equals(name) 식의 equals 사용은 피해야 합니다.
o나 o.getName()이 null인 경우에,
null.equals(..) -> NullPointerException이 발생하거든요.

이 경우에는 o의 null 여부 검증 후
name.equals(o.getName())을 하는게 좋겠죠.

상당히 많이 발생하는 에러케이스라 기억해두시면 좋을 것 같아요.
"A.equals(B) 할 때 A가 null이 아니라는 것이 보장되어야 한다."

jjj0611 replied:
null pointer exception...
조심해야겠습니다!

dain7672 replied:
네네. 이부분은 확실히!

dain7672 commented on src/main/java/domain/Car.java:
```diff
@@ -1,36 +1,62 @@
+        return this.position == position;
+    }
+
+    public CarSnapShot getCarSnapShot() {
```
car.getCarSnapShot()
vs.
CarSnapShot.from(car)

어떤 방식이 좋을까요? (함수를 어디에 위치시키는 게 좋을까요?)

jjj0611 replied:
자동차 자체가 내 기록 이거야! 하고 주는게 맞다고 생각했는데,
CarSnapShot을 모르게 하고, 밖에서 기록을 해주는게 더 좋은 방식인 것 같아 수정했습니다!

dain7672 replied:
네. 정답은 없고, CarSnapShot 클래스의 정체성을 어떻게 바라보느냐의 차이일 것 같아요.
저의 경우에는 CarSnapShot이 화면 출력을 위한 것이라는 점에서 MVC 모델, 또는 layered architecture에서의 view, presentation layer에 해당한다고 생각했어요. (DTO)

실제 운영 상황에서는 CarSnapShot 뿐만 아니라 어드민을 위한 DTO나, Car의 특정 필드만 담고있는 DTO도 필요할 수도 있는데요.

해당 DTO를 생성하는 메소드는 DTO 자신이 가지는 것이 좋다고 봅니다.

1. 클래스 간의 참조관계를 생각했을 때,
view, 또는 presentation layer에서의 요구사항을 model이나 domain에서 가지게 되는 경우에는
어떤 DTO의 변경을 위해 Car 클래스를 변경한 것이 다른 DTO에도 영향을 끼치게 될 수도 있구요.

2. 역할과 책임을 생각했을 때
Car는 자동차에 대한 핵심 비즈니스 로직, CarSnapShot은 Car을 보여주기 위한 로직을 가지고 있는 클래스라는 점에서도 해당 클래스의 생성 메소드를 자기가 가지게 하는 것이 좋은 것 같아 보여요.

dain7672 commented on src/main/java/domain/RacingResult.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private int getRound(CarSnapShots carSnapShots) {
+        return carSnapShotsList.indexOf(carSnapShots) + DIFFERENCE_BETWEEN_INDEX_ROUND;
```
for (Snapshot snapshot : list) 에서
list.indexOf(snapshot) 으로 index를 찾아낼 거라면

for (int i=0;i< ...) 방식을 사용하는 게 나아보여요.

dain7672 replied:
아니면,
CarSnapShot 클래스에
int round 멤버 변수를 가지게 할 수도 있겠죠?

jjj0611 replied:
List에 들어 있으면 index로서 round의 역할을 할 수 있는데,
이때 round라는 변수를 가지고 있는게 혹시 중복은 아닌가요?

일단 for(int i=0.. 식으로 변경했습니다!

dain7672 replied:
CarSnapShots가 list에 들어있지 않는 경우 또는, 순서가 보장되어 있지 않은 경우에 필요할 것 같습니다.ㅎㅎ
말씀하신대로 현재 상황에서는 list의 index로 충분한 것 같습니다~

dain7672 commented on src/main/java/domain/CarSnapShots.java:
```diff
@@ -0,0 +1,55 @@
+        this.carSnapShots = carSnapShots;
+    }
+
+    public void printCarSnapShots() {
```
이름은 print로도 충분할 것 같아요~

jjj0611 replied:
CarSnapShots.print()와 같이 호출하면 되니 그렇겠네요....!

dain7672 commented on src/main/java/game/GameInputScanner.java:
```diff
@@ -0,0 +1,21 @@
+package game;
+
+import java.util.Scanner;
+
+public class GameInputScanner {
+
+    private static final Scanner scanner = new Scanner(System.in);
+
+    private static final String CAR_NAME_ASKING_STATEMENT = "경주할 자동차 이름을 입력하세요. 이름은 쉼표(,) 단위로 구분 됩니다.";
+    private static final String TRIAL_ASKING_STATEMENT = "시도 횟수를 입력하세요.";
+
+    public static String getCarNameFromUser() {
+        System.out.println(CAR_NAME_ASKING_STATEMENT);
+        return scanner.nextLine();
+    }
+
+    public static String getTrial() {
+        System.out.println(TRIAL_ASKING_STATEMENT);
+        return scanner.nextLine();
+    }
+}
```
👍

dain7672 commented on src/main/java/domain/RacingCars.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RacingCars {
```
일급 컬렉션을 잘 사용해주셨네요.
RacingCars 라고 이름붙일 때는 자동차들을 담는 용도의 정체성을 가진 (일급 컬렉션) 클래스일 거라고 유추하게 되는 것 같아요.

그런데 지금과 같은 경우에는,
해당 클래스에서 데이터의 포장 뿐만 아니라 경주도 직접 진행하고 있는데요.
그렇다면 CarRacing, Racing의 이름을 가진 "레이싱 경주"를 위한 정체성을 가진 클래스가 되는 게 맞아 보여요.

jjj0611 replied:
아 알겠습니다.
RacingCars와 같은 경우 그냥 자동차들만 들고 있겠네 처럼 생각하기 때문에 헷갈릴 수 있겠네요.
수정 했습니다.

dain7672 reviewed (commented):
재주 선생님 반갑습니다!
코드리뷰를 맡게된 김다인이에요.

"이해하기 쉽고, 정확하게 의미를 전달하려면?" 의 관점으로 리뷰를 진행해봤어요.

지금 진행하는 리뷰는 오류를 방지하기 위한 것이라기보다는
"유지보수"하기 좋은 코드에 대해 논의하기 위해서라고 생각해요.

객체지향과 클린코드의 궁극적인 목표는 "코드의 변경"이 용이하게 이루어지기 위한 것이라고 볼 수 있는데요,
나의 동료나, 미래의 내가 다시 코드를 보고, 변경된 요구사항을 반영하거나 할 때에 드는 비용을 최소화하기 위한 것이지요. 미래에 변경사항이 없을 예정이라면, 좋은 코드를 작성할 필요도 없겠죠.

코드를 작성하실 때에 위와 같은 고민을 계속해보시는 것을 추천드려요!

좋은 커밋 관리도 위와 관련이 있어요.
커밋 메시지에 feature 번호를 달아주신 건 아주 좋아요.
나중에 확인하기 용이하거든요.
다만 중요하다고 여겨지는 변경사항들이 뒤로갈수록 커밋 메시지에 많이 빠져있는 건 아쉬웠어요.
간단하게라도 변경사항을 모두 커버하는 메시지를 남길 수 있도록 노력해주시면 좋을 것 같아요.

마지막으로는 사견입니다만, 오픈소스 작업이나 글로벌 기업에서 일을 하는 것이 아니라면, 한국어로 커밋메시지를 작성하는 게 나을 수도 있지 않을까 싶어요. 뭐든 간에 중요한 것은 "이걸 읽는 사람들이 잘 이해하려면?" 이거든요.ㅎㅎ

감사합니다~

jjj0611 committed:
.gitignore에 DS_Store 추가

jjj0611 committed:
Car 클래스의 eqauls 실행시 NullPointerException을 방지하기 위해 코드 변경

jjj0611 committed:
RacingCar test 리팩토링

	- 자동차를 생성하는 테스트
	- 자동차 생성시 중복된 차량을 허용하지 않는 테스트
	- 자동차 생성시 단일한 경주를 허용하지 않는 테스트
	- 세 테스트를 분리

jjj0611 committed:
racingcar 게임 리팩토링

	- CarTest에서 예외 확인하는 테스트 분리
	- Car 클래스에서 전진하는 메서드 변경
	- CarSnapshot을 생성하는 방식 변경
	- CarSnapshots를 print하는 메서드명을 변경
	- Result 출력하는 반복문 방식을 변경

kwangilcho commented:
👍

dain7672 commented:
👍
